### ISSUE TITLE
Pin change interrupts (or other async inputs which can generate interrupts)

### ISSUE THREAD

**Issue by dhylands** [micropython/micropython#183]
So I'd like to figure out how to deal with pin change interrupts.

The way I deal with them in C is that I use lockless queue between the ISR layer and the C layer.
https://github.com/dhylands/projects/blob/master/common/CBUF.h
In order to be lockless it needs to have a length which is a power of 2.

The ISR queues up an event (which would probably include some sort of pin identifier, type of event, and a timestamp), and the foreground polls the queue periodically.

In an ideal world, I'd like to hook this into the select mechanism somehow (along with serial ports and anything that can generate input).

So basically, I'm soliciting feedback/ideas on how we should deal with this.

**Comment by dpgeorge** [comment]
This thread discussed some important things but is old and can be closed.  The current state of things is:
- Python code can be called on an interrupt, but it has restriction (mainly no memory allocation)
- Python code can be scheduled on an interrupt to run "soon", so it can allocate memory
- A callback for a pin change is registered via pin.irq()
